sycl/include/sycl/detail/accessor_iterator.hpp

```diff
@@ -0,0 +1,375 @@
+//==------------ accessor_iterator.hpp - SYCL standard header file ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/id.hpp>
+
+#include <cstddef>
+#include <iterator>
+#include <ostream>
+#include <type_traits>
+
+/// \file accessor_iterator.hpp
+/// The file contains implementation of accessor iterator class.
+///
+/// The reason why we can't use a plain pointer as an interator and have to
+/// implement a custom class here is explained in section 4.7.6.8. Ranged
+/// accessors of SYCL 2020 specification. A couple of quotes from there:
+///
+/// > Accessors of type accessor and host_accessor can be constructed from a
+/// > sub-range of a buffer by providing a range and offset to the constructor.
+/// >
+/// > If the ranged accessor is multi-dimensional, the sub-range is allowed to
+/// > describe a region of memory in the underlying buffer that is not
+/// > contiguous in the linear address space.
+/// >
+/// > Most of the accessor member functions which provide a reference to the
+/// > underlying buffer elements are affected by a ranged accessor’s offset and
+/// > range. ... In addition, the accessor’s iterator functions iterate only
+/// > over the elements that are within the sub-range.
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+
+template <typename _AccessorDataT, int _AccessorDimensions,
+          access::mode _AccessMode, access::target _AccessTarget,
+          access::placeholder _IsPlaceholder, typename _PropertyListT>
+class accessor;
+
+namespace detail {
+
+template <typename _DataT, int _Dimensions> class __accessor_iterator {
+public:
+  using difference_type = std::ptrdiff_t;
+  using value_type = _DataT;
+  // FIXME: this should likely include address space
+  using pointer = _DataT *;
+  using reference = _DataT &;
+  using iterator_category = std::random_access_iterator_tag;
+
+  __accessor_iterator() = default;
+
+  reference operator*() {
+    return *(_MDataPtr + __get_absolute_offset_to_buffer());
+  }
+
+  __accessor_iterator &operator++() {
+    if (_MLinearId < _MEnd)
```

Comment:
What is the reason for having all these tests?
I guess that it will diminish the performance.

Reply:
> What is the reason for having all these tests?

See my reply to your other comment. Basically, the reasoning is quite weak and we should probably indeed remove those checks.

Reply:
Simplified the implementation in 8e2db4ef3f3a6c95c161a5b4d0825ce1a3173fa0